DESIGN.md

```diff
@@ -325,18 +325,99 @@ In the addition of the previous rules, an operator for `Flowable`:
 
 ### Creation
 
-Creation of the various types should be exposed through factory methods that provide safe construction.
+Unlike RxJava 1.x, 2.x base classes are abstract and stateless and generally no longer wrap an `OnSubscribe` callback - this saves allocation in assembly time without limiting the expressiveness. Operator methods and standard factories still live as final on the base classes. 
+
+Since Java still doesn't have extension methods, "adding" more operators can only happen through helper methods such as `lift(C -> C)` and `compose(R -> P)` where `C` is the default consumer type (i.e., `rs.Subscriber`), `R` is the base type (i.e., `Flowable`) and `P` is the base interface (i.e., `rs.Publisher`). As before, the library itself may gain or lose standard operators and/or overloads through the same community process.
```

Comment:
Okay, I would rather only have one method for "extending" functionality? Is that also your goal? In which case we should discuss the merits of compose as a replacement for extend. 

For comparison I'm including the signature of extend.

```java
<O2, X extends Consumable<O2>> X extend(Function<Consumer<O>, X> f)
```

Which in the case of a Flowable would be as follows (note that I have replaced the `Consumable<O2>` with `Publisher<U>`)

```java
<O2, X extends<Publisher<U>> X extend(Function<Consumer<Subscriber<? super T>>, X> f)
```

So the only substantial difference between compose and extend is that with compose the function consumes a concrete Flowable type and with extend it consumes an OnSubscribe. 

I think my concern is that the `compose` functions _must_ know the absolute type of the thing that you are converting from (or at least a base class/interface). This makes it more restrictive than what `extend` allows, converting from any variation of a `Flowable` by virtue of unwrapping and accessing the internal callback. So composition functions would be case specific, and would not be reusable for different types of flowable, observable, single, or completable. 

Note that I am **NOT** proposing making conversion functions from a flowable that also work on an observable for instance. What I am proposing is that if you have different variations of an Observable that the same logic could be use to convert it to a different thing. I think that there is the potential for better code reuse with `extend`. Also my intent is that extending functions for the core RxJava types (i.e. the `Flowable`, `Observable`, `Single`, `Completable` and all of there variants which implement their respective `Consumable<S>` for their subscriber/observer type) be written and packaged with RxJava. But this would not be limited to just internal apis as the pattern could be useful for libraries which write their own `Consumable`.